File: demos/demo_ParIS_DTW.py
import sys
import os
import numpy as np
import tempfile

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

try:
    from daisy import DistanceType, ParIS
except ImportError:
    print("Error: ParIS is not yet available in Python bindings.")
    print("ParIS requires file-based data and needs to be added to pybinds/setup.cpp")
    print("See demo_ParIS_DTW.cpp for the C++ implementation.")
    sys.exit(1)

def main():
    
    n_database = 200000
    dim = 96
    n_query = 10
    k = 5

    np.random.seed(100)   
    db = np.random.randn(n_database, dim).astype(np.float32)

    np.random.seed(50)
    query = np.random.randn(n_query, dim).astype(np.float32)

    print(f"Loaded {n_database} database points and {n_query} query points with dimension {dim}")

    temp_file = tempfile.NamedTemporaryFile(delete=False, suffix='.bin')
    temp_db_file = temp_file.name
    temp_file.close()
    
    try:
        
        db.tofile(temp_db_file)
        print(f"Database written to temporary file: {temp_db_file}")

        index = ParIS(DistanceType.DTW)
        index.setNumThreads(4)

        warp_window = max(1, int(dim * 0.1))
        index.setWarpingWindow(warp_window)
        print(f"Warping window set to: {warp_window}")

        index.buildIndex(temp_db_file, dim, n_database)
        print(">>> Finished indexing")

        I, D = index.searchIndex(query, k)
        print(">>> Finished search")

        for query_num in range(n_query):
            print(f"Query {query_num}:")
            print("Distances:", D[query_num])
            print("Indices:", I[query_num])
            print()

    finally:
        
        if os.path.exists(temp_db_file):
            os.remove(temp_db_file)
            print(f"Cleaned up temporary file: {temp_db_file}")

if __name__ == "__main__":
    main()

File: demos/demo_ParIS_DTW.cpp
#include "../commons/dataloaders.hpp"
#include "../lib/daisy.hpp"
#include <cstdio>
#include <cstring>
#include <string>

int main(){
    
    daisy::idx_t n_database = 200000;
    unsigned long long dim = 96;
    unsigned long long n_query = 10;
    daisy::idx_t k = 5;

    float *database = loadRandomData(n_database, dim, 100, true);
    float *query = loadRandomData(n_query, dim, 50, true);

    printf("Loaded %llu database points and %llu query points with dimension %llu\n", n_database, n_query, dim);

    std::string temp_db_file = "/tmp/paris_test_db.bin";
    FILE *fp = fopen(temp_db_file.c_str(), "wb");
    if (fp == nullptr) {
        fprintf(stderr, "Error: Could not create temporary database file\n");
        delete[] database;
        delete[] query;
        return 1;
    }
    fwrite(database, sizeof(float), n_database * dim, fp);
    fclose(fp);

    daisy::ParIS paris_search(daisy::DistanceType::DTW);
    paris_search.setNumThreads(4);

    int warp_window = std::max(1, static_cast<int>(dim * 0.1));
    paris_search.setWarpingWindow(warp_window);

    paris_search.buildIndex(temp_db_file, dim, n_database);

    daisy::idx_t *I = new daisy::idx_t[n_query * k];
    float *D = new float[n_query * k];
    paris_search.searchIndex(query, n_query, k, I, D);

    for (daisy::idx_t i = 0; i < n_query; i++)
    {
        printf("Query %llu: ", i);
        for (daisy::idx_t j = 0; j < k; j++)
        {
            printf("%llu ", I[i * k + j]);
        }
        printf("\n");
    }

    delete[] database;
    delete[] query;
    delete[] I;
    delete[] D;
    remove(temp_db_file.c_str());  

    return 0;
}
